pkg/services/vmoperator/control_plane_endpoint.go

```diff
@@ -0,0 +1,226 @@
+func (s CPService) ReconcileControlPlaneEndpointService(ctx *vmware.ClusterContext, netProvider services.NetworkProvider) (*clusterv1.APIEndpoint, error) {
+	ctx.Logger.V(4).Info("Reconciling control plane VirtualMachineService for cluster")
+
+	// If the NetworkProvider does not support a load balancer, this should be a no-op
+	if !netProvider.HasLoadBalancer() {
+		return nil, nil
+	}
+
+	vmService, err := s.getVMControlPlaneService(ctx)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			err = errors.Wrapf(err, "failed to check if VirtualMachineService exists")
+			conditions.MarkFalse(ctx.VSphereCluster, infrav1.LoadBalancerReadyCondition, infrav1.LoadBalancerCreationFailedReason, clusterv1.ConditionSeverityWarning, err.Error())
+			return nil, err
+		}
+
+		// Get the provider annotations for the ControlPlane Service.
+		annotations, err := netProvider.GetVMServiceAnnotations(ctx)
+		if err != nil {
+			err = errors.Wrapf(err, "failed to get provider VirtualMachineService annotations")
+			conditions.MarkFalse(ctx.VSphereCluster, infrav1.LoadBalancerReadyCondition, infrav1.LoadBalancerCreationFailedReason, clusterv1.ConditionSeverityWarning, err.Error())
+			return nil, err
+		}
+
+		vmService, err = s.createVMControlPlaneService(ctx, annotations)
+		if err != nil {
+			err = errors.Wrapf(err, "failed to create VirtualMachineService")
+			conditions.MarkFalse(ctx.VSphereCluster, infrav1.LoadBalancerReadyCondition, infrav1.LoadBalancerCreationFailedReason, clusterv1.ConditionSeverityWarning, err.Error())
+			return nil, err
+		}
+	}
+
+	// See if the LB has a VIP assigned, and delay reconciliation until it does
+	vip, err := getVMServiceVIP(vmService)
+	if err != nil {
+		err = errors.Wrapf(err, "VirtualMachineService LB does not yet have VIP assigned")
+		conditions.MarkFalse(ctx.VSphereCluster, infrav1.LoadBalancerReadyCondition, infrav1.WaitingForLoadBalancerIPReason, clusterv1.ConditionSeverityInfo, err.Error())
+		return nil, err
+	}
+
+	cpEndpoint, err := getAPIEndpointFromVIP(vmService, vip)
+	if err != nil {
+		err = errors.Wrapf(err, "VirtualMachineService LB does not have an apiserver endpoint")
+		conditions.MarkFalse(ctx.VSphereCluster, infrav1.LoadBalancerReadyCondition, infrav1.WaitingForLoadBalancerIPReason, clusterv1.ConditionSeverityWarning, err.Error())
+		return nil, err
+	}
+
+	conditions.MarkTrue(ctx.VSphereCluster, infrav1.LoadBalancerReadyCondition)
+	return cpEndpoint, nil
+}
+
+func controlPlaneVMServiceName(ctx *vmware.ClusterContext) string {
+	return fmt.Sprintf("%s-control-plane-service", ctx.Cluster.Name)
+}
+
+// ClusterRoleVMLabels returns labels applied to a VirtualMachine in the cluster. The Control Plane
+// VM Service uses these labels to select VMs, as does the Cloud Provider.
+func clusterRoleVMLabels(ctx *vmware.ClusterContext, controlPlane bool) map[string]string {
+	result := map[string]string{
+		clusterSelectorKey: ctx.Cluster.Name,
```

Comment:
labels updated below should reflect then here and then in the `VirtualMachineService`'s selector + `VirtualMachine` labels, since this is called when calling `getVMLabels()`